### PR TITLE
issue: 1074708 Add TSO capability

### PR DESCRIPTION
**Pull request by igor-ivanov**
This pull request is based on previous as https://github.com/Mellanox/libvma/pull/447
enables TSO functionality.

### CODE REVIEW

File: src/vma/proto/dst_entry.cpp [hunk 1]
```diff
@@ -664,6 +676,14 @@ void dst_entry::do_ring_migration(lock_base& socket_lock, resource_allocation_ke
 
 	ring* old_ring = m_p_ring;
 	m_p_ring = new_ring;
+	if (m_sge) {
```
Comment (liranoz12) [inline src/vma/proto/dst_entry.cpp]: Any reason to have a different value ? i am not sure...

File: src/vma/dev/ring_tap.h [hunk 2]
```diff
@@ -69,8 +69,9 @@ class ring_tap : public ring_slave
 
 	virtual int modify_ratelimit(struct vma_rate_limit_t &rate_limit) { NOT_IN_USE(rate_limit); return 0; }
 	void inc_cq_moderation_stats(size_t sz_data) { NOT_IN_USE(sz_data); }
-	virtual int get_max_tx_inline() { return 0; }
 	virtual uint32_t get_underly_qpn() { return -1; }
+	virtual uint32_t get_max_inline_data() { return 0; }
+	virtual uint32_t get_max_send_sge() { return 1; }
```
Comment (liranoz12) [inline src/vma/dev/ring_tap.h]: Why do we want to disable it for ring tap   ?

File: src/vma/proto/dst_entry.cpp [hunk 3]
```diff
@@ -307,6 +311,14 @@ bool dst_entry::resolve_ring()
 			m_p_ring = m_p_net_dev_val->reserve_ring(m_ring_alloc_logic.create_new_key(m_pkt_src_ip));
 		}
 		if (m_p_ring) {
+			if (m_sge) {
```
Comment (liranoz12) [inline src/vma/proto/dst_entry.cpp]: Do we expect m_sge to be not-null ?

File: src/vma/dev/ring.h [hunk 4]
```diff
@@ -112,8 +112,11 @@ class ring
 	ring_user_id_t		generate_id() { return 0; };
 	virtual ring_user_id_t	generate_id(const address_t src_mac, const address_t dst_mac, uint16_t eth_proto, uint16_t encap_proto, uint32_t src_ip, uint32_t dst_ip, uint16_t src_port, uint16_t dst_port) = 0;
 	virtual int		modify_ratelimit(struct vma_rate_limit_t &rate_limit) = 0;
-        virtual uint32_t        get_max_inline_data() = 0;
-        virtual uint32_t        get_max_send_sge(void) = 0;
+        virtual uint32_t	get_max_inline_data() = 0;
+        virtual uint32_t	get_max_send_sge(void) = 0;
+        virtual uint32_t	get_max_payload_sz(void) = 0;
+        virtual uint16_t	get_max_header_sz(void) = 0;
+        virtual bool		is_tso(void) = 0;
```
Comment (liranoz12) [inline src/vma/dev/ring.h]: Please consider using a single function instead of those 5 -> get_tso_caps()

File: src/vma/proto/dst_entry_tcp.cpp [hunk 5]
```diff
@@ -107,13 +109,26 @@ ssize_t dst_entry_tcp::fast_send(const iovec* p_iov, const ssize_t sz_iov, bool
...
-		m_p_send_wqe = (total_packet_len < m_max_inline ? &m_inline_send_wqe : &m_not_inline_send_wqe);
-
-		m_sge[0].addr = (uintptr_t)((uint8_t*)p_pkt + hdr_alignment_diff);
-		m_sge[0].length = total_packet_len;
-		m_p_send_wqe->wr_id = (uintptr_t)p_tcp_iov[0].p_desc;
-		p_tcp_iov[0].p_desc->tx.p_ip_h = &p_pkt->hdr.m_ip_hdr;
-		p_tcp_iov[0].p_desc->tx.p_tcp_h =(struct tcphdr*)((uint8_t*)(&(p_pkt->hdr.m_ip_hdr))+sizeof(p_pkt->hdr.m_ip_hdr));
+		if ((total_packet_len >= m_max_inline) && m_p_ring->is_tso()) {
+			send_wqe_h.init_not_inline_wqe(send_wqe, m_sge, 1);
+			send_wqe_h.enable_tso(send_wqe,
```
Comment (liranoz12) [inline src/vma/proto/dst_entry_tcp.cpp]: Can we introduce another send-wqe element pre-configured for tso ?

File: src/vma/proto/dst_entry_tcp.cpp [hunk 6]
```diff
@@ -107,13 +109,26 @@ ssize_t dst_entry_tcp::fast_send(const iovec* p_iov, const ssize_t sz_iov, bool
 		/* update L3(Total Length) with total size of L3 header, TCP header and data */
 		p_pkt->hdr.m_ip_hdr.tot_len = (htons)(p_tcp_iov[0].iovec.iov_len + m_header.m_ip_header_len);
 
-		m_p_send_wqe = (total_packet_len < m_max_inline ? &m_inline_send_wqe : &m_not_inline_send_wqe);
-
-		m_sge[0].addr = (uintptr_t)((uint8_t*)p_pkt + hdr_alignment_diff);
-		m_sge[0].length = total_packet_len;
-		m_p_send_wqe->wr_id = (uintptr_t)p_tcp_iov[0].p_desc;
-		p_tcp_iov[0].p_desc->tx.p_ip_h = &p_pkt->hdr.m_ip_hdr;
-		p_tcp_iov[0].p_desc->tx.p_tcp_h =(struct tcphdr*)((uint8_t*)(&(p_pkt->hdr.m_ip_hdr))+sizeof(p_pkt->hdr.m_ip_hdr));
```
Comment (liranoz12) [inline src/vma/proto/dst_entry_tcp.cpp]: We will have issues when tx csum is not supported.

File: config/m4/verbs.m4 [hunk 7]
```diff
@@ -195,6 +195,15 @@ if test "x$vma_cv_verbs" == x2; then
...
 
+    AC_CHECK_DECL([IBV_EXP_DEVICE_ATTR_TSO_CAPS],
+        [], [], [[#include <infiniband/verbs_exp.h>]])
+    AC_MSG_CHECKING([for tso support])
+    AS_IF([test "x$ac_cv_have_decl_IBV_EXP_DEVICE_ATTR_TSO_CAPS" = "xyes" -a "x$vma_cv_directverbs" == x2],
+        [AC_DEFINE_UNQUOTED([HAVE_TSO], [1], [Define to 1 if tso is supported])]
+        [AC_MSG_RESULT([yes])],
+        [AC_MSG_RESULT([no])]
+    )
+
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply config/m4/verbs.m4]: it is not implemented

File: src/vma/dev/ring_simple.cpp [hunk 8]
```diff
@@ -250,10 +250,31 @@ void ring_simple::create_resources()
...
+#ifdef HAVE_TSO
+	if (m_p_ib_ctx->get_ibv_device_attr()->comp_mask & IBV_EXP_DEVICE_ATTR_TSO_CAPS) {
+		const struct ibv_exp_tso_caps *caps = &m_p_ib_ctx->get_ibv_device_attr()->tso_caps;
+		if (ibv_is_qpt_supported(caps->supported_qpts, IBV_QPT_RAW_PACKET) ||
+			ibv_is_qpt_supported(caps->supported_qpts, IBV_QPT_UD)) {
+			m_tso.max_payload_sz = caps->max_tso;
+			/* ETH(14) + IP(20) + TCP(20) + TCP OPTIONS(40) */
+			m_tso.max_header_sz = 94;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_simple.cpp]: looks reasonable

File: src/vma/dev/ring_simple.h [hunk 9]
```diff
@@ -198,6 +211,17 @@ class ring_simple : public ring_slave
...
+
+	/* Maximum number of scatter/gather elements */
+	int m_max_sge;
+
+	struct {
+		/* Maximum length of TCP payload for TSO */
+		uint32_t max_payload_sz;
+
+		/* Maximum length of header for TSO */
+		uint16_t max_header_sz;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/ring_simple.h]: qp_mgr

File: src/vma/dev/ring_tap.h (hunk 2 above)
Comment (igor-ivanov) [reply src/vma/dev/ring_tap.h]: probably compilation isssue

File: src/vma/proto/dst_entry_tcp.cpp (hunk 6 above)
Comment (igor-ivanov) [reply src/vma/proto/dst_entry_tcp.cpp]: can you explain.

**Pull request merged**